Introduce broadcast API for event sharing

gwbaik9717 opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it?
This PR implements a broadcast API, which enables the sharing of a broader range of general events beyond the current document and presence events in Yorkie's Publish-Subscribe model.

#### Any background context you want to provide?
**1. Broadcast Events:**

Users can now broadcast custom events with a specified topic and payload.
The payload can be of any type, as long as it is serializable.

```javascript
// Broadcast an event with a topic and payload
const payload = 'hello';
doc.broadcast('TOPIC_NAME', payload);
```

**2. Subscribe to Broadcast Events:**

Users can subscribe to specific topics and handle the events via a callback function.
The callback is triggered whenever an event with the corresponding topic is broadcast.

```javascript
// Subscribe to a specific topic for broadcast events
doc.subscribe('broadcast', ({value: {topic, payload, clientID}}) => {
    // Handle the broadcast event for the specified topic
});
```

#### What are the relevant tickets?
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Related to https://github.com/yorkie-team/yorkie/issues/628

### Checklist
- [x] Added relevant tests or not required
- [x] Didn't break anything


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Introduced a `broadcast` method in the Client class to send messages to specified topics.
	- Added new event types `Broadcast` and `LocalBroadcast` to enable clients to subscribe to broadcast messages.
	- Implemented a method to validate the serializability of payloads for broadcasting.
	- Enhanced the Document class with a `subscribe` method for clients to register handlers for broadcast events.

- **Bug Fixes**
	- Improved error handling for broadcasting unserializable payloads within the Client class.

- **Tests**
	- Added test cases to verify successful broadcasting of serializable payloads and error handling for invalid payloads.
	- Increased test coverage for event subscriptions related to broadcast events.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gwbaik9717 commented:
@sejongk 

There are a few things I'd like to share:

**1. How Handlers for Each Topic Are Managed:**

Inside the Document class, I've added a `broadcastEventHandlers` which is a map to manage the handlers for each topic.

```typescript
// document.ts 
export class Document<T, P extends Indexable = Indexable> {
  // skipped

  private broadcastEventHandlers: Map<
    string,
    (topic: string, payload: any) => void
  >;
}
```

When a user subscribes to a specific topic for broadcast events:

```typescript
// User subscribes to a broadcast event with the topic "TOPIC_NAME"
doc.subscribeBroadcastEvent('TOPIC_NAME', (topic, payload) => {
    // Handle the broadcast event for the specified topic
});
```

The handler is registered in the broadcastEventHandlers map:

```typescript
export class Document<T, P extends Indexable = Indexable> {
  
  public subscribeBroadcastEvent(
    topic: string,
    handler: (topic: string, payload: any) => void,
    error?: ErrorFn,
  ): Unsubscribe {
  
    // Register the handler in broadcastEventHandlers 
    this.broadcastEventHandlers.set(topic, handler);
```

**2. How `PbDocEventType` is Converted to `DocEventType`:**

Within the `applyWatchStream` method of the `Document` class, which applies the given watch stream response to the document, the conversion from `PbDocEventType` to `DocEventType` occurs as follows:

```javascript
  public applyWatchStream(resp: WatchDocumentResponse) {
      // skipped
      } else if (type === PbDocEventType.DOCUMENT_BROADCAST) {
        if (resp.body.value.body) {
          const { topic, payload } = resp.body.value.body;
          const decoder = new TextDecoder();

          event.push({
            type: DocEventType.Broadcast,
            value: { topic, payload: JSON.parse(decoder.decode(payload)) },
          });
        }
      }

      if (event.length > 0) {
        this.publish(event);
      }
    }
  }
```

**3. Test Case for Sending Unserializable Payloads:**

As discussed earlier, I've added a test case in `client_test.ts` to handle unserializable payloads. If there are any additional test cases needed, please let me know.


Lastly, I've verified that the broadcast API works as expected in several examples, but I believe further testing is necessary to ensure complete coverage.

sejongk commented on packages/sdk/test/integration/client_test.ts:
```diff
@@ -863,4 +863,24 @@ describe.sequential('Client', function () {
     }, task.name);
   });
+
+  it('Throw error when broadcasting unserializeable payload', async ({
```
How about adding a test for a success case?

gwbaik9717 replied:
Thank you for the suggestion. I've added a test case for successful broadcasting. 
However, I also wanted to add the following test case, which tests broadcasting between two clients. Unfortunately, the test code below doesn't work as expected because both `c1` and `c2`, provided by `withTwoClientsAndDocuments`, share the same `rpcAddr` . I believe this prevents them from behaving like separate clients in the test environment. Is there any way to make the following test case work?

```javascript
  it('Successfully broadcast serializeable payload', async ({ task }) => {
    await withTwoClientsAndDocuments<{ t: Text }>(async (c1, d1, c2, d2) => {
      const spy = vi.fn();

      const topic = 'test';
      const unsub = d2.subscribeBroadcastEvent(topic, spy);

      const payload = { a: 1, b: '2' };
      await c1.broadcast(d1, topic, payload);

      expect(spy.mock.calls.length).toBe(1);
      expect(spy.mock.calls[0][0]).toBe(topic);
      expect(spy.mock.calls[0][1]).toEqual(payload);

      unsub();
    }, task.name);
  });

```

gwbaik9717 replied:
@sejongk 
Thank you, it works now! I tried both of the approaches you suggested. However, the first approach using `.sync()` didn't seem to work due to the uncertainty in the timing of when the broadcast event would arrive. So, I decided to go with the second approach. Instead of creating a new `withTwoClientsAndDocuments` function with `SyncMode.Realtime`, I added a new optional `syncMode` parameter. Here is the new function signature.

```javascript
async function withTwoClientsAndDocuments<T>(
  callback: (
    c1: Client,
    d1: Document<T>,
    c2: Client,
    d2: Document<T>,
  ) => Promise<void>,
  title: string,
  syncMode: SyncMode = SyncMode.Manual,
): Promise<void>
```

Additionally, I added a few more test cases to handle subscribing and unsubscribing.

chacha912 replied:
It seems that broadcast events are propagated through watchDocument, so it appears that testing isn't possible when real-time sync is not enabled, as the watchDocument stream isn't connected. In the Go implementation, the watch interface was exposed to users, allowing for event verification.(ref. https://github.com/yorkie-team/yorkie/issues/584) 

For future documentation, it would be helpful to clarify that broadcast events can only be received when sync mode is set to real-time.

gwbaik9717 replied:
Oh, now I see why it didn't work before. I opened an issue regarding documentation [issue](https://github.com/yorkie-team/yorkie-team.github.io/issues/164), and will make sure to include the following limitation: Broadcast events can only be received when the sync mode is set to real-time.

sejongk commented on packages/sdk/src/document/document.ts:
```diff
@@ -1027,6 +1049,36 @@ export class Document<T, P extends Indexable = Indexable> {
+   * on the document. The callback will be called when the document receives a
+   * broadcast event with the given topic.
+   */
+  public subscribeBroadcastEvent(
```
@hackerwins, @chacha912  I think we need to discuss whether to integrate this function with the `subscribe` function above by passing an event type (e.g., 'my-presence', 'broadcast') and a broadcast topic (or null) together (defining a new type using `pair`).

chacha912 replied:
I'm still unsure about the best approach for the subscribe interface. I've summarized the interfaces of other libraries and our current proposal:

* [Liveblocks](https://liveblocks.io/docs/api-reference/liveblocks-client#Room.subscribe.event)
   * Doesn't use separate topics, filters by event type within the event
   * The callback function receives `{event, user, connectionId}` as its argument, which includes information about the user who sent the event.

```js
// User 1
room.broadcastEvent({ type: "REACTION", emoji: "🔥" });

// User 2
const unsubscribe = room.subscribe("event", ({ event, user, connectionId }) => {
  //                                                  ^^^^ Will be User 1
  if (event.type === "REACTION") {
    // Do something
  }
});
```

* [Supabase](https://supabase.com/docs/guides/realtime/broadcast)
   * Callback function receives only `(payload)` as an argument
   * Subscribe to based on event topic.

```js
// User1
const channelB = supabase.channel('room-1')
channelB.send({
  type: 'broadcast',
  event: 'test',
  payload: { message: 'hello, world' },
})

// User2
const channelA = supabase.channel('room-1')
channelA
  .on(
    'broadcast',
    { event: 'test' }, // topic
    (payload) => messageReceived(payload)
  )
  .subscribe()
```

* Yorkie
   * Option 1
      ```js
      // User1
      const payload = 'hello';
      doc.broadcast('TOPIC_NAME', payload);
      
      // User2
      doc2.subscribeBroadcastEvent('TOPIC_NAME', (topic, payload) => {
          // Handle the broadcast event for the specified topic
      });
      ```
   * Option 2 (current)
      ```js
      // User1
      const payload = 'hello';
      doc.broadcast('TOPIC_NAME', payload);
      
      // User2
      doc2.subscribe({ type: 'broadcast', topic: 'TOPIC_NAME'}, (topic, payload) => {
          // Handle the broadcast event for the specified topic
      });
      ```

* Is there a reason for passing both `topic` and `payload` in the callback function?
* While Option 2 uses a common subscribe method, I'm concerned that adding the `BroadcastSubscribePair` type might have made it more complex.

gwbaik9717 replied:
@chacha912 @sejongk @hackerwins 

I totally agree that `BroadcastSubscribePair` might be too complex to use. What do you think about aligning it with the existing event subscription pattern? The main reason I introduced `BroadcastSubscribePair` was to prevent users from having to compare topic values manually. However, after reviewing the subscription methods for other events, I noticed that users are already handling similar comparisons themselves.

```javascript
doc.subscribe('presence', (event) => {
              // Skipped
              } else if (event.type === 'unwatched') {
                cursors.removeCursor(event.value.clientID);
              } 
});
```

Therefore, I think it might be better to unify the subscription approach for broadcast events in the same way:

```javascript
 const broadcastTopic = 'test';
 const unsubscribe = d2.subscribe('broadcast', (topic, payload) => {
          if (topic === broadcastTopic) {
            eventCollector.add(payload);
          }
});
```

Of course, the potential issue with the interface mentioned above is that handlers could be triggered unintentionally if the comparison logic isn't implemented correctly. However, it also provides users with more flexibility. I'll go ahead and implement the new subscription interface in a commit—please review if it is okay.

sejongk commented on packages/sdk/src/client/client.ts:
```diff
@@ -584,6 +585,59 @@ export class Client {
+  /**
+   * `broadcast` broadcasts the given payload to the given topic.
+   */
+  public broadcast<T, P extends Indexable>(
```
I think we should use an interface like `doc.broadcast(topic, payload)` rather than `client.broadcast(doc, topic, payload)`, as it aligns with how a client subscribes to a broadcast event using `doc.subscribe(topic, handler)`. Without removing or changing the existing function, it seems possible to add another broadcast function in `document.ts` that calls this function.

gwbaik9717 replied:
@chacha912 
Thank you for pointing this out!  However, I haven't yet figured out how to call a method in `Client` directly from the document. Is there any way to get an instance of `Client` by `actorId`? 

```javascript
// document.ts
  public broadcast(topic: string, payload: any): void {

    const actorId = this.changeID.getActorID()
    
    // have to get `Client` instance by actorId  in order to call method `broadcast`
    const client =  getClientByActorId(actorId) // may be something like this
    
```

sejongk replied:
I think one possible approach is to add some attribute members (e.g., `client`) to the `Document` class for this. However, I'm not sure if this is the best idea 😅.

gwbaik9717 replied:
@hackerwins @chacha912 
Based on what you've mentioned, when attaching a document using the client's `attach` method, we could probably set client reference, within the `Document` class. However, this also needs to be discussed as well.

chacha912 replied:
I agree that the `doc.broadcast()` interface seems more appropriate. This has led to the addition of a `client` property within the `Document` class. At present, I can't think of a better alternative approach for now. (cc. @hackerwins )

sejongk reviewed (changes requested):
Overall, it looks good to me 👍. 
I've left some comments.

hackerwins commented:
@gwbaik9717 
CI failed due to incorrect pnpm version in CI configuration. It is currently fixed in the main branch.

gwbaik9717 commented:
@sejongk 

Here is an update.

**1. Unified Broadcast Event Subscription**

Users can now use the existing `subscribe` method to subscribe to broadcast events. Since subscribing to a broadcast event requires both a `type` and a `topic`, I've introduced a new `SubscribePair` type, which is structured as follows:

```javascript
type SubscribePair = {
  type: string;
};
```

Building on this, the `BroadcastSubscribePair` extends `SubscribePair`:
```javascript
type BroadcastSubscribePair = {
  type: 'broadcast';
  topic: string;
} & SubscribePair;
```

Here’s an example of how to subscribe to a broadcast event:
```javascript
   const unsubscribe = d2.subscribe(
        { type: 'broadcast', topic: "TOPIC_NAME" },
        (topic, payload) => {},
   );
```

**2. Call `broadcast` from `Document` not `Client`**

Broadcasting is now handled directly by the `Document` not `Client`, aligning with the subscription mechanism.

To support this, I’ve added a `client` attribute within the `Document` class. This `client` attribute is an instance of the `Client` and is set when the client attaches to the document, and unset when it detaches.

```javascript
export class Document<T, P extends Indexable = Indexable> {
  // skipped
  private client?: Client;
  
    /**
   * `setClient` sets the client of this document.
   *
   * @internal
   */
  public setClient(client?: Client): void {
    this.client = client;
  }

}
```

Here’s an example of how to broadcast a topic with payload.

```javascript
// must be serializable
const payload = {a:1, b:"2"}
await doc.broadcast("TOPIC_NAME", payload);
```

sejongk reviewed (approved):
LGTM 👍.
@hackerwins Is there anything else to review?

hackerwins commented:
@sejongk

Sorry for the late review. I'll check it today.

CC) @chacha912, @gwbaik9717

gwbaik9717 commented on packages/sdk/test/integration/client_test.ts:
```diff
@@ -863,4 +863,137 @@ describe.sequential('Client', function () {
+      const payload = { a: 1, b: '2' };
+      await d1.broadcast(broadcastTopic, payload);
+
+      // Assuming that every subscriber can receive the broadcast event within 1000ms.
+      await new Promise((res) => setTimeout(res, 1000));
```
@chacha912 
Thank you for the suggestion. While `EventCollector` is a great alternative to timeouts and I've successfully applied it in several test cases, some scenarios were not well-suited for its use.

For example, in the test case below, the handler won't be called because it hasn't subscribed to the topic. As a result, using `EventCollector` wouldn't be effective here:

```javascript
it('Should not trigger the handler for an unsubscribed broadcast event', async ({
  task,
}) => {
  await withTwoClientsAndDocuments<{ t: Text }>(
    async (c1, d1, c2, d2) => {
      const spy = vi.fn();

      const broadcastTopic1 = 'test1';
      const broadcastTopic2 = 'test2';

      const unsubscribe = d2.subscribe(
        { type: 'broadcast', topic: broadcastTopic2 },
        () => {
           // wont'be called
           spy()
        },
      );

      const payload = { a: 1, b: '2' };
      await d1.broadcast(broadcastTopic1, payload);

      // Assuming that every subscriber can receive the broadcast event within 1000ms.
      await new Promise((res) => setTimeout(res, 1000));

      expect(spy.mock.calls.length).toBe(0);

      unsubscribe();
    },
    task.name,
    SyncMode.Realtime,
  );
});
```

chacha912 replied:
While we can use `eventCollector.getLength()` to check if no events have occurred, it seems we might need an additional method to test that no events occur during a timeout period. For now, I think it's fine to proceed with the approach you've suggested. Thank you for making those modifications.

gwbaik9717 commented on packages/sdk/src/document/document.ts:
```diff
@@ -589,6 +617,17 @@ export class Document<T, P extends Indexable = Indexable> {
    */
   private isUpdating: boolean;
 
+  /**
+   * `broadcastEventHandlers` is a map of broadcast event handlers.
+   * The key is the topic of the broadcast event, and the value is the handler.
+   */
+  private broadcastEventHandlers: Map<
+    string,
+    DocEventCallbackMap<P>['broadcast']
+  >;
```
Thank you for pointing this out. You were absolutely right—there was no need to add `broadcastEventHandlers`. I've referenced the GO-SDK, which includes `broadcastEventHandlers`, leading to my confusion. I've since removed it and confirmed that everything works correctly.

chacha912 commented:
@hackerwins, I've reviewed, but could you please check it once more in case I missed anything? (Please check the comments I've left with cc regarding the subscribe interface and client property)

gwbaik9717 commented on packages/sdk/src/document/document.ts:
```diff
@@ -1970,4 +2078,18 @@ export class Document<T, P extends Indexable = Indexable> {
+  /**
+   * `broadcast` the payload to the given topic.
+   */
+  public broadcast(topic: string, payload: any): Promise<void> {
```
I'll address this issue in the upcoming issue [#591](https://github.com/yorkie-team/yorkie-js-sdk/issues/591) 😀

gwbaik9717 commented:
@hackerwins @chacha912 @sejongk 

Here's an update.

In order to resolve the circular reference issue, we’ve removed the `client` reference from the `Document` class.  Instead, we've decided to have `Client` subscribe to `Document`'s broadcast events and handle them by calling `Client`'s `broadcast` method. 

However, this solution presents two challenges:

**1. Unsubscribing Handlers**

We had to ensure that handlers registered in `subscribe` are properly unsubscribed when document is detached from client. 
To address this, we’ve decided to manage `unsubscribeBroadcastEvent` within the `Attachment` class, since the lifecycle of the client’s attachment and the sub/unsub process are the same.

```javascript
export class Attachment<T, P extends Indexable> {
  // skipped
  unsubscribeBroadcastEvent: Unsubscribe;
}
```

With this approach, when attaching a document to a client using the `attach` method of Client, we subscribe to the document’s event stream, and we `unsubscribe` when detaching:

```javascript
  public attach<T, P extends Indexable>(
    doc: Document<T, P>,
    options: {
      initialPresence?: P;
      syncMode?: SyncMode;
    } = {},
  ): Promise<Document<T, P>> {
    // skipped
    const unsubscribeBroacastEvent = doc.subscribe(
      'broadcast',
      (topic, payload, onBroadcastError) => {
        try {
          this.broadcast(doc.getKey(), topic, payload);
        } catch (e: unknown) {
          if (e instanceof Error) {
            onBroadcastError?.(e);
          }
        }
      },
    );
    
    this.attachmentMap.set(
          doc.getKey(),
          new Attachment(
            this.reconnectStreamDelay,
            doc,
            res.documentId,
            syncMode,
            unsubscribeBroacastEvent,
         ),
    );
```

```javascript
  private detachInternal(docKey: DocumentKey) {
    // skipped
    attachment.unsubscribeBroadcastEvent();
```

**2. Error Handling for Broadcast Events**

The second challenge involves handling errors during broadcast events, such as when broadcasting an unserializable payload. Initially, we could simply use a try-catch block to handle errors when directly accessing the client's broadcast method:

```javascript
try {
   doc.broadcast(broadcastTopic, payload)
}catch(e){
 // do something
}
```

However, since we’re now dealing with an event stream, a different approach to error handling is required. To accommodate this, I've introduced a new interface for broadcasting events. This interface includes an error handling function that triggers when a broadcast event fails.

```javascript
// Usage
const errorFn = () => {
    // Do something
}
doc.broadcast("topic", payload, errorFn)
```

gwbaik9717 commented:
Here's an update.

As mentioned earlier, we've decided to have the `Client` subscribe to the `Document`'s broadcast events and handle them by invoking the `Client`'s broadcast method. The challenge I encountered was distinguishing between `local` and `remote` broadcast events. Since the client is now subscribed to the document's broadcast events, both `local` and `remote` events are received, which creates the following issue:

<img width="862" alt="image" src="https://github.com/user-attachments/assets/8b668b69-3682-4f2b-b89a-62c1b5cc9b4c">

Without differentiation, this setup can lead to an unexpected infinite loop. This happens because the client, upon receiving its own broadcast event, will rebroadcast it, creating a never-ending cycle.

<img width="862" alt="image" src="https://github.com/user-attachments/assets/93c29968-7348-432b-aa3e-6f88423b8259">

This scenario can be reproduced with the following test code. The test simulates a situation where the publisher is broadcasting an event and, at the same time, subscribing to the broadcast event. Currently, since Yorkie does not support self-broadcast filtering, the subscribe handler gets called, leading to numerous recursive broadcasts, as evidenced by the test results:

```javascript
 it('Should not trigger the handler for a broadcast event sent by the publisher to itself', async ({
    task,
  }) => {
    await withTwoClientsAndDocuments<{ t: Text }>(
      async (c1, d1, c2, d2) => {
        const eventCollector1 = new EventCollector<[string, any]>();
        const eventCollector2 = new EventCollector<[string, any]>();
        const broadcastTopic = 'test';
        const payload = { a: 1, b: '2' };

        // Publisher subscribes to the broadcast event
        const unsubscribe1 = d1.subscribe('broadcast', (topic, payload) => {
          if (topic === broadcastTopic) {
            eventCollector1.add([topic, payload]);
          }
        });

        const unsubscribe2 = d2.subscribe('broadcast', (topic, payload) => {
          if (topic === broadcastTopic) {
            eventCollector2.add([topic, payload]);
          }
        });

        d1.broadcast(broadcastTopic, payload);

        // Assuming that subscribers can receive the broadcast event within 1000ms.
        await new Promise((res) => setTimeout(res, 1000));

        unsubscribe1();
        unsubscribe2();

        assert.equal(eventCollector1.getLength(), 0);
        assert.equal(eventCollector2.getLength(), 1);
      },
      task.name,
      SyncMode.Realtime,
    );
  });
```

<img width="862" alt="Screenshot 2024-09-01 at 12 00 34 PM" src="https://github.com/user-attachments/assets/633f8c97-c908-43b2-90e8-49180b129fef">

&nbsp;

To resolve this issue, I've introduced a `LocalBroadcast` event type in addition to the existing `Broadcast` event. The `LocalBroadcast` event is only triggered when the local document attempts to broadcast. This addition allows us to clearly distinguish between `local` and `remote` broadcast events, preventing the infinite loop scenario by ensuring that a client does not handle its own broadcast events.
&nbsp;

```javascript
export enum DocEventType {
  /**
   * `Broadcast` means that the broadcast event is received from the remote client.
   */
  Broadcast = 'broadcast',

  /**
   * `LocalBroadcast` means that the broadcast event is sent from the local client.
   */
  LocalBroadcast = 'local-broadcast',
}
```

```javascript
 /**
   * `broadcast` the payload to the given topic.
   */
  public broadcast(topic: string, payload: any, error?: ErrorFn) {
    const broadcastEvent: LocalBroadcastEvent = {
      type: DocEventType.LocalBroadcast,
      value: { topic, payload },
      error,
    };

    this.publish([broadcastEvent]);
  }
```

hackerwins reviewed (approved):
@gwbaik9717 @sejongk @chacha912 
Thanks for your contribution and review.
